c/rc_lock.c

```diff
@@ -0,0 +1,462 @@
+  uint8_t flags = args_bytes_seg.ptr[0];
+  CHECK2(flags == IdentityFlagsPubkeyHash || flags == IdentityFlagsOwnerLock ||
+             flags == IdentityFlagsPubkeyHashRc ||
+             flags == IdentityFlagsOwnerLockRc,
+         ERROR_INVALID_MOL_FORMAT);
+
+  CHECK2(args_bytes_seg.size >= (1 + BLAKE160_SIZE + BLAKE2B_BLOCK_SIZE),
+         ERROR_ARGUMENTS_LEN);
+  args->id.flags = args_bytes_seg.ptr[0];
+  memcpy(args->id.blake160, args_bytes_seg.ptr + 1, BLAKE160_SIZE);
+  memcpy(args->rc_root, args_bytes_seg.ptr + 1 + BLAKE160_SIZE,
+         sizeof(args->rc_root));
+
+exit:
+  return err;
+}
+
+int load_and_hash_witness(blake2b_state *ctx, size_t start, size_t index,
+                          size_t source, bool hash_length) {
+  uint8_t temp[ONE_BATCH_SIZE];
+  uint64_t len = ONE_BATCH_SIZE;
+  int ret = ckb_load_witness(temp, &len, start, index, source);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (hash_length) {
+    blake2b_update(ctx, (char *)&len, sizeof(uint64_t));
+  }
+  uint64_t offset = (len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : len;
+  blake2b_update(ctx, temp, offset);
+  while (offset < len) {
+    uint64_t current_len = ONE_BATCH_SIZE;
+    ret = ckb_load_witness(temp, &current_len, start + offset, index, source);
+    if (ret != CKB_SUCCESS) {
+      return ret;
+    }
+    uint64_t current_read =
+        (current_len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : current_len;
+    blake2b_update(ctx, temp, current_read);
+    offset += current_read;
+  }
+  return CKB_SUCCESS;
+}
+
+int verify_secp256k1_blake160_sighash_all(uint8_t *pubkey_hash,
+                                          uint8_t *signature_bytes) {
+  int ret;
+  uint64_t len = 0;
+  unsigned char temp[MAX_WITNESS_SIZE];
+  uint64_t read_len = MAX_WITNESS_SIZE;
+  uint64_t witness_len = MAX_WITNESS_SIZE;
+
+  /* Load witness of first input */
+  ret = ckb_load_witness(temp, &read_len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  witness_len = read_len;
+  if (read_len > MAX_WITNESS_SIZE) {
+    read_len = MAX_WITNESS_SIZE;
+  }
+
+  /* load signature */
+  mol_seg_t lock_bytes_seg;
+  ret = extract_witness_lock(temp, read_len, &lock_bytes_seg);
+  if (ret != 0) {
+    return ERROR_ENCODING;
+  }
+  if (lock_bytes_seg.size < SIGNATURE_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
+  }
+  /* Load tx hash */
+  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
+  len = BLAKE2B_BLOCK_SIZE;
+  ret = ckb_load_tx_hash(tx_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len != BLAKE2B_BLOCK_SIZE) {
+    return ERROR_SYSCALL;
+  }
+
+  /* Prepare sign message */
+  unsigned char message[BLAKE2B_BLOCK_SIZE];
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+
+  /* Clear lock field to zero, then digest the first witness */
+  memset((void *)lock_bytes_seg.ptr, 0, lock_bytes_seg.size);
```

Comment:
This is far too tricky, I think it's better we leave a note saying `lock_bytes_seg.ptr` actually points to memory in `temp` buffer, otherwise we will for sure be confused by this code later.